pkg/controller/core.oam.dev/v1alpha2/application/dispatcher.go

```diff
@@ -0,0 +1,215 @@
+func ByTraitType(readyTraits, checkTraits []*unstructured.Unstructured) TraitFilter {
+	generateFn := func(manifests []*unstructured.Unstructured) map[string]bool {
+		out := map[string]bool{}
+		for _, obj := range manifests {
+			out[obj.GetLabels()[oam.TraitTypeLabel]] = true
+		}
+		return out
+	}
+	readyMap := generateFn(readyTraits)
+	checkMap := generateFn(checkTraits)
+	return func(trait appfile.Trait) bool {
+		return !checkMap[trait.Name] && readyMap[trait.Name]
+	}
+}
+
+// manifestDispatcher is a manifest dispatcher
+type manifestDispatcher struct {
+	run         func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision, clusterName string) (bool, error)
+	healthCheck func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision) (bool, error)
+}
+
+func (h *AppHandler) generateDispatcher(appRev *v1beta1.ApplicationRevision, readyWorkload *unstructured.Unstructured, readyTraits []*unstructured.Unstructured, overrideNamespace string) ([]*manifestDispatcher, error) {
+	dispatcherGenerator := func(options DispatchOptions) *manifestDispatcher {
+		assembleManifestFn := func(skipApplyWorkload bool) (bool, []*unstructured.Unstructured) {
+			manifests := options.Traits
+			skipWorkload := skipApplyWorkload || options.Workload == nil
+			if !skipWorkload {
+				manifests = append([]*unstructured.Unstructured{options.Workload}, options.Traits...)
+			}
+			return skipWorkload, manifests
+		}
+
+		dispatcher := new(manifestDispatcher)
+		dispatcher.healthCheck = func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision) (bool, error) {
+			skipWorkload, manifests := assembleManifestFn(wl.SkipApplyWorkload)
+			if !h.resourceKeeper.ContainsResources(manifests) {
+				return false, nil
+			}
+			_, isHealth, err := h.collectHealthStatus(ctx, wl, appRev, options.OverrideNamespace, skipWorkload,
+				ByTraitType(readyTraits, options.Traits))
+			if err != nil {
+				return false, err
+			}
+			return isHealth, nil
+		}
+		dispatcher.run = func(ctx context.Context, wl *appfile.Workload, appRev *v1beta1.ApplicationRevision, clusterName string) (bool, error) {
+			skipWorkload, dispatchManifests := assembleManifestFn(wl.SkipApplyWorkload)
+			if isHealth, err := dispatcher.healthCheck(ctx, wl, appRev); !isHealth || err != nil {
+				if err := h.Dispatch(ctx, clusterName, common.WorkflowResourceCreator, dispatchManifests...); err != nil {
+					return false, errors.WithMessage(err, "Dispatch")
+				}
+				status, isHealth, err := h.collectHealthStatus(ctx, wl, appRev, options.OverrideNamespace, skipWorkload,
+					ByTraitType(readyTraits, options.Traits))
+				if err != nil {
+					return false, errors.WithMessage(err, "CollectHealthStatus")
+				}
+				if options.Stage < DefaultDispatch {
+					status.Healthy = false
+					if status.Message == "" {
+						status.Message = "waiting for previous stage healthy"
```

Comment:
why you don't check the `isHealth` and set the message directly?

Reply:
The `isHealth` here is only for the current resource. I don't think it is reasonable to directly set the  status that should belong to the workload. For example, the health status of the workload should be maintained during the post phase.

Reply:
OK, that's fine